learntools/machine_learning/nbs/tut4.ipynb

```diff
@@ -0,0 +1,192 @@
+    "mean_absolute_error(y, predicted_home_prices)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "_cell_guid": "030a4791-a484-48b8-85e5-e5ab94cc07c6",
+    "_uuid": "07a8221afe2b026623b0534a43379a1084010c8b"
+   },
+   "source": [
+    "# The Problem with \"In-Sample\" Scores\n",
+    "\n",
+    "The measure we just computed can be called an \"in-sample\" score. We used a single \"sample\" of houses for both building the model and evaluating it. Here's why this is bad.\n",
+    "\n",
+    "Imagine that, in the large real estate market, door color is unrelated to home price. \n",
+    "\n",
+    "However, in the sample of data you used to build the model, all homes with green doors were very expensive. The model's job is to find patterns that predict home prices, so it will see this pattern, and it will always predict high prices for homes with green doors.\n",
+    "\n",
+    "Since this pattern was derived from the training data, the model will appear accurate in the training data.\n",
+    "\n",
+    "But if this pattern doesn't hold when the model sees new data, the model would be very inaccurate when used in practice.\n",
+    "\n",
+    "Since models' practical value come from making predictions on new data, we measure performance on data that wasn't used to build the model. The most straightforward way to do this is to exclude some data from the model-building process, and then use those to test the model's accuracy on data it hasn't seen before. This data is called **validation data**.\n",
+    "\n",
+    "\n",
+    "# Coding It\n",
+    "\n",
+    "\n",
+    "The scikit-learn library has a function `train_test_split` to break up the data into two pieces. We'll use some of that data as training data to fit the model, and we'll use the other data as validation data to calculate `mean_absolute_error`.\n",
+    "\n",
+    "Here is the code:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "_cell_guid": "b9d1a3b5-dd2e-49ab-8902-64e350b025e6",
+    "_uuid": "0b1782e8adc52855dc1168cf0ad5214c19a30330",
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "from sklearn.model_selection import train_test_split\n",
+    "\n",
+    "# split data into training and validation data, for both predictors and target\n",
+    "# The split is based on a random number generator. Supplying a numeric value to\n",
+    "# the random_state argument guarantees we get the same split every time we\n",
+    "# run this script.\n",
+    "train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)\n",
+    "# Define model\n",
+    "melbourne_model = DecisionTreeRegressor()\n",
+    "# Fit model\n",
+    "melbourne_model.fit(train_X, train_y)\n",
+    "\n",
+    "# get predicted prices on validation data\n",
+    "val_predictions = melbourne_model.predict(val_X)\n",
+    "print(mean_absolute_error(val_y, val_predictions))"
```

Comment:
We kind of breeze past the result of this experiment, but I think it's really worth remarking on. Going from being off by $435 on average to being off by $263,519 is an incredible difference, and a dramatic demonstration of why validating on training data is so dangerous.

It's such a huge difference, it makes me really curious to learn more. How did our model overfit our data so precisely? What does our decision tree actually look like? Presumably it's really deep? How do the number of leaves compare to the number of training instances? What features is it overfitting to? (i.e. what's the analogy to the "green door" in our case?) Maybe this would be worth investigating a bit in the next tutorial (underfitting and overfitting)?

Reply:
Also (sorry, I can't stop thinking about this), an average error of $263k seems *bad*. Is it? What MAE do we get by just guessing the overall training-set average price every time? Is this heavily influenced by outliers?

I like that the exercise after this has us look at a few examples of predictions on the validation data and compare them to the actual price. Could be interesting to do that here to get a feel for the kinds of errors we're dealing with.